### PR TITLE
fix issue 38716 insert_all enum values by correctly type casting the …

### LINKED ISSUE #38716: `insert_all` does not work properly with enums as it does not cast types
### Steps to reproduce

```ruby
class BulkInsertItem < ActiveRecord::Base
    enum enum_value: {
     value_A: 1
    }
end

ActiveRecord::Schema.define do
  create_table :bulk_insert_items, force: true do |t|
    t.integer :enum_value, null: false
  end
end

BulkInsertItem.insert_all!([
   { enum_value: "value_A" }
]) # this results in in ability to insert enum value, as `string` tries to be converted to `integer`

BulkInsertItem.insert_all!([
   { enum_value: 1 }
]) # it requires to performing manual conversion, as `enum` decorates an attribute type
```

### Expected behavior

That the convenient form does work properly for `insert_all`:

```ruby
BulkInsertItem.insert_all!([
   { enum_value: "value_A" }
])
```

### The problem

It appears that problem lies in https://github.com/rails/rails/blob/master/activerecord/lib/active_record/insert_all.rb#L184.

We serialise data using a type as declared by the DB schema,
but we completely ignore `decorate_attribute_type` there.

It appears that the line should rather be:

```ruby
keys.index_with { |key| type_for_attribute(columns[key]) }
```

This would return a decorator type, or database type and allow to properly serialise
it's representation, in the same way as `.save!` is doing.

### System configuration
**Rails version**: `6.0.2`

### PR DESCRIPTION
**Pull request by yoelblum**
fix issue 38716 insert_all enum values by correctly type casting the attributes

### CODE REVIEW

**Comment by simi** [comment]
:+1: I'll check locally some edge-cases and performance.

File: activerecord/test/cases/insert_all_test.rb [hunk 1]
```diff
@@ -335,6 +335,11 @@ def test_upsert_all_works_with_partitioned_indexes
                  Measurement.where(city_id: 2).pluck(:logdate, :peaktemp, :unitsales)
   end
 
+  def test_insert_all_with_enum_values
+    Book.insert_all [{ status: :published, isbn: "1234566" }, { status: :proposed, isbn: "1234567" }]
+    assert Book.where(isbn: ["1234566", "1234567"]).pluck(:status), [:published, :proposed]
```
Comment (eugeneius) [inline activerecord/test/cases/insert_all_test.rb]:
```suggestion
    assert_equal ["published", "proposed"], Book.where(isbn: ["1234566", "1234567"]).pluck(:status)
```

`assert` passes if its first argument is truthy, which will always be the case here - this test as written can't fail.

Comment (yoelblum) [reply activerecord/test/cases/insert_all_test.rb]: @eugeneius you're 100% right

File: activerecord/test/cases/insert_all_test.rb [hunk 2]
```diff
@@ -335,6 +335,11 @@ def test_upsert_all_works_with_partitioned_indexes
                  Measurement.where(city_id: 2).pluck(:logdate, :peaktemp, :unitsales)
   end
 
+  def test_insert_all_with_enum_values
+    Book.insert_all [{ status: :published, isbn: "1234566" }, { status: :proposed, isbn: "1234567" }]
+    assert_equal Book.where(isbn: ["1234566", "1234567"]).pluck(:status), ['published', 'proposed']
```
Comment (kamipo) [inline activerecord/test/cases/insert_all_test.rb]: I'm not sure why rubocop did not run, but we use double quotes style for string literal, and use `assert_equal expected, actual` style like as other assertions.
And also, the result of `pluck` without `order` is not always stable order (especially on PostgreSQL), add `order(:id)`.

```suggestion
    assert_equal ["published", "proposed"], Book.where(isbn: ["1234566", "1234567"]).order(:id).pluck(:status)
```

**Comment by ayufan** [comment]
Yay. Thank you :)

**Comment by rishav-csenitjsr** [comment]
Nice. Was doing the same thing.
Let’s say my class was Book which was having enum_value, But when I was printing Book.all after insert_all command, I was getting nil value for enum_value. Btw, thanks for sharing the PR

**Comment by rishav-csenitjsr** [comment]
@yoelblum specs are failing. Can you please look into it ?

**Pull request merged**